File: killer_sudoku/solver/engine/rules/locked_candidates.py
"""R10b LockedCandidates — digit in a unit confined to one cage or box.

Two patterns (both fired from a ROW, COL, or BOX trigger):

  Unit → Cage  (Cage-Line Reduction):
    When all cells in a row/col/box that carry digit d lie within a single
    cage, d is locked to the unit-cage intersection.  Since d must appear
    somewhere in the row/col/box, and all those positions are inside the
    cage, d can be eliminated from cage cells outside the unit.

  Unit → Box  (Box-Line Reduction):
    When all cells in a row or column that carry digit d lie within a single
    3×3 box, d is locked to the row/col-box intersection.  It can therefore
    be eliminated from the other cells of that box outside the row/column.

These are the unit-first counterparts to CageIntersection (cage-first) and
PointingPairs (box-first).

Fires on COUNT_DECREASED for ROW, COL, and BOX units.
"""

from __future__ import annotations

import dataclasses
from collections.abc import Iterator

from killer_sudoku.solver.engine.board_state import BoardState
from killer_sudoku.solver.engine.hint import HintResult
from killer_sudoku.solver.engine.rule import RuleContext
from killer_sudoku.solver.engine.rules._labels import cell_label, unit_label
from killer_sudoku.solver.engine.rules._registry import hintable_rule
from killer_sudoku.solver.engine.types import (
    Cell,
    Elimination,
    RuleResult,
    Trigger,
    Unit,
    UnitKind,
)


@dataclasses.dataclass(frozen=True)
class _Match:
    """One firing of LockedCandidates."""

    digit: int
    source_unit: Unit  # the row/col/box that triggered
    carriers: frozenset[Cell]  # cells in source_unit that carry d
    target_unit: Unit  # the cage or box receiving the eliminations
    pattern: str  # "unit_cage" | "unit_box"
    eliminations: list[Elimination]


@hintable_rule
class LockedCandidates:
    """R10b: digit in a unit confined to one cage or box — eliminate from container."""

    name = "LockedCandidates"
    priority = 11
    triggers: frozenset[Trigger] = frozenset({Trigger.COUNT_DECREASED})
    unit_kinds: frozenset[UnitKind] = frozenset(
        {UnitKind.ROW, UnitKind.COL, UnitKind.BOX}
    )

    # ── internal helpers ────────────────────────────────────────────────────

    @staticmethod
    def _iter_matches(ctx: RuleContext) -> Iterator[_Match]:
        """Yield one _Match per (digit, target-unit) pair that fires."""
        assert ctx.unit is not None
        board: BoardState = ctx.board
        unit_cells = ctx.unit.cells
        unit_kind = ctx.unit.kind

        for d in range(1, 10):
            carriers = [(r, c) for r, c in unit_cells if d in board.candidates[r][c]]
            if len(carriers) < 2:
                continue

            # ── Pattern 1: Unit → Cage (Cage-Line Reduction) ─────────────────
            common_cage_ids: set[int] | None = None
            for r, c in carriers:
                cell_cages = {
                    uid
                    for uid in board.cell_unit_ids(r, c)
                    if board.units[uid].kind == UnitKind.CAGE
                }
                if common_cage_ids is None:
                    common_cage_ids = cell_cages
                else:
                    common_cage_ids &= cell_cages
                if not common_cage_ids:
                    break

            if common_cage_ids:
                for cage_uid in common_cage_ids:
                    elims = [
                        Elimination(cell=(r, c), digit=d)
                        for r, c in board.units[cage_uid].cells
                        if (r, c) not in unit_cells and d in board.candidates[r][c]
                    ]
                    if elims:
                        yield _Match(
                            digit=d,
                            source_unit=ctx.unit,
                            carriers=frozenset(carriers),
                            target_unit=board.units[cage_uid],
                            pattern="unit_cage",
                            eliminations=elims,
                        )

            # ── Pattern 2: Unit → Box (Box-Line Reduction) ────────────────────
            if unit_kind in (UnitKind.ROW, UnitKind.COL):
                rows = {r for r, _ in carriers}
                cols = {c for _, c in carriers}
                box_rows = {r // 3 for r in rows}
                box_cols = {c // 3 for c in cols}
                if len(box_rows) == 1 and len(box_cols) == 1:
                    br, bc = next(iter(box_rows)), next(iter(box_cols))
                    box_uid = board.box_unit_id(br * 3, bc * 3)
                    elims = [
                        Elimination(cell=(r, c), digit=d)
                        for r, c in board.units[box_uid].cells
                        if (r, c) not in unit_cells and d in board.candidates[r][c]
                    ]
                    if elims:
                        yield _Match(
                            digit=d,
                            source_unit=ctx.unit,
                            carriers=frozenset(carriers),
                            target_unit=board.units[box_uid],
                            pattern="unit_box",
                            eliminations=elims,
                        )

    @staticmethod
    def _build_hint(m: _Match) -> HintResult:
        """Construct a HintResult from a confirmed _Match."""
        d = m.digit
        carriers_str = ", ".join(sorted(cell_label(c) for c in m.carriers))
        source_lbl = unit_label(m.source_unit)
        target_lbl = unit_label(m.target_unit)
        elim_cells_str = ", ".join(sorted(cell_label(e.cell) for e in m.eliminations))

        if m.pattern == "unit_cage":
            explanation = (
                f"In {source_lbl}, {d} can only go in {carriers_str}. "
                f"All those cells belong to the same cage ({target_lbl}'s cage). "
                f"Since {d} must appear somewhere in {source_lbl} and all its "
                f"candidates are inside that cage, {d} can be eliminated from "
                f"{elim_cells_str} (cage cells outside {source_lbl})."
            )
            display_name = "Locked Candidates (Cage-Line)"
        else:
            explanation = (
                f"In {source_lbl}, {d} can only go in {carriers_str}. "
                f"All those cells lie within {target_lbl}. "
                f"Since {d} must appear somewhere in {source_lbl} and all its "
                f"candidates are locked to {target_lbl}, {d} can be eliminated "
                f"from {elim_cells_str} "
                f"(the other cells in {target_lbl} outside {source_lbl})."
            )
            display_name = "Locked Candidates (Box-Line)"

        return HintResult(
            rule_name="LockedCandidates",
            display_name=display_name,
            explanation=explanation,
            highlight_cells=m.carriers | {e.cell for e in m.eliminations},
            eliminations=m.eliminations,
        )

    # ── SolverRule protocol ─────────────────────────────────────────────────

    def apply(self, ctx: RuleContext) -> RuleResult:
        """Eliminate d from any container that holds all of this unit's d-candidates."""
        seen: set[tuple[Cell, int]] = set()
        elims: list[Elimination] = []
        for m in self._iter_matches(ctx):
            for e in m.eliminations:
                key = (e.cell, e.digit)
                if key not in seen:
                    seen.add(key)
                    elims.append(e)
        return RuleResult(eliminations=elims)

    # ── Hint interface ──────────────────────────────────────────────────────

    def as_hints(
        self, ctx: RuleContext, eliminations: list[Elimination]
    ) -> list[HintResult]:
        """Return one HintResult per (digit, target-unit) match."""
        if not eliminations:
            return []
        hints: list[HintResult] = []
        seen: set[tuple[Cell, int]] = set()
        for m in self._iter_matches(ctx):
            new_keys = {(e.cell, e.digit) for e in m.eliminations} - seen
            if new_keys:
                seen.update(new_keys)
                hints.append(self._build_hint(m))
        return hints

File: web/src/engine/rules/lockedCandidates.ts
/**
 * LockedCandidates — R10b: digit in a unit confined to one cage or box.
 *
 * Mirrors Python's `killer_sudoku.solver.engine.rules.locked_candidates` module.
 *
 * Two patterns:
 *  - unit→cage (cage-line reduction): row/col/box d-candidates all in same cage → eliminate from cage outside unit
 *  - unit→box (box-line reduction): row/col d-candidates all in same box → eliminate from box outside row/col
 */

import type { HintResult } from '../hint.js';
import type { RuleContext } from '../rule.js';
import { Cell, Elimination, emptyResult, RuleResult, Trigger, UnitKind } from '../types.js';
import { cellLabel, unitLabel } from './_labels.js';

interface _Match {
  digit: number;
  sourceUnitId: number;
  carriers: Cell[];
  targetUnitId: number;
  pattern: 'unit_cage' | 'unit_box';
  eliminations: Elimination[];
}

export class LockedCandidates {
  readonly name = 'LockedCandidates';
  readonly killerOnly = false;
  readonly displayName = 'Locked Candidates';
  readonly description = `\
Locked Candidates — a digit confined to the intersection of two units can be eliminated from the remainder of each unit.

Two patterns:

Cage-line (unit → cage): If all candidates for d in a row, column, or box lie within one cage, then d must be placed inside that cage's intersection with the unit. The cage cannot place d outside the intersection, so d is eliminated from cage cells outside the unit.

Box-line (row/col → box): If all candidates for d in a row or column lie within one 3×3 box, then d is locked to that row/col-box intersection. The box cannot place d outside the intersection, so d is eliminated from box cells outside the row or column.

Proof (box-line, one case, exhaustive because all candidates are inside the box):
  All d-candidates in row R lie in box B → d goes in B∩R → box already has d covered → d ≠ any cell in B outside R.

Guards:
  carriers.length >= 2        at least two candidates needed (one would be a Hidden Single)
  commonCageIds?.size > 0     cage-line: all carriers share a common cage
  boxRows.size === 1 && boxCols.size === 1   box-line: all carriers share one box`.trim();
  readonly priority = 14;
  readonly triggers: ReadonlySet<Trigger> = new Set([Trigger.COUNT_DECREASED]);
  readonly unitKinds: ReadonlySet<UnitKind> = new Set([UnitKind.ROW, UnitKind.COL, UnitKind.BOX]);

  private _iterMatches(ctx: RuleContext): _Match[] {
    if (!ctx.unit) return [];
    const board = ctx.board;
    const unitCells = ctx.unit.cells as Cell[];
    const unitCellSet = new Set(unitCells.map(([r,c]) => `${r},${c}`));
    const matches: _Match[] = [];

    for (let d = 1; d <= 9; d++) {
      const carriers = unitCells.filter(([r, c]) => board.cands(r, c).has(d));
      if (carriers.length < 2) continue;

      // Pattern 1: unit → cage
      let commonCageIds: Set<number> | null = null;
      for (const [r, c] of carriers) {
        const cageUids = board.cellUnitIds(r, c).filter((uid: number) => board.units[uid]!.kind === UnitKind.CAGE);
        const cellCages = new Set<number>(cageUids);
        commonCageIds = commonCageIds === null ? cellCages : new Set<number>([...commonCageIds].filter((uid: number) => cellCages.has(uid)));
        if (!commonCageIds.size) break;
      }
      if (commonCageIds?.size) {
        for (const cageUid of commonCageIds) {
          const elims = (board.units[cageUid]!.cells as Cell[])
            .filter(([r,c]) => !unitCellSet.has(`${r},${c}`) && board.cands(r, c).has(d))
            .map(cell => ({ cell, digit: d }));
          if (elims.length) matches.push({ digit: d, sourceUnitId: ctx.unit.unitId, carriers, targetUnitId: cageUid, pattern: 'unit_cage', eliminations: elims });
        }
      }

      // Pattern 2: row/col → box
      if (ctx.unit.kind === UnitKind.ROW || ctx.unit.kind === UnitKind.COL) {
        const boxRows = new Set(carriers.map(([r]) => r / 3 | 0));
        const boxCols = new Set(carriers.map(([,c]) => c / 3 | 0));
        if (boxRows.size === 1 && boxCols.size === 1) {
          const br = [...boxRows][0]!, bc = [...boxCols][0]!;
          const boxUid = board.boxUnitId(br * 3, bc * 3);
          const elims = (board.units[boxUid]!.cells as Cell[])
            .filter(([r,c]) => !unitCellSet.has(`${r},${c}`) && board.cands(r, c).has(d))
            .map(cell => ({ cell, digit: d }));
          if (elims.length) matches.push({ digit: d, sourceUnitId: ctx.unit.unitId, carriers, targetUnitId: boxUid, pattern: 'unit_box', eliminations: elims });
        }
      }
    }
    return matches;
  }

  apply(ctx: RuleContext): RuleResult {
    const seen = new Set<string>();
    const elims: Elimination[] = [];
    for (const m of this._iterMatches(ctx)) {
      for (const e of m.eliminations) {
        const key = `${e.cell[0]},${e.cell[1]}:${e.digit}`;
        if (!seen.has(key)) { seen.add(key); elims.push(e); }
      }
    }
    return { ...emptyResult(), eliminations: elims };
  }

  asHints(ctx: RuleContext, eliminations: Elimination[]): HintResult[] {
    if (!eliminations.length) return [];
    const seen = new Set<string>();
    const hints: HintResult[] = [];
    for (const m of this._iterMatches(ctx)) {
      const newElims = m.eliminations.filter(e => {
        const k = `${e.cell[0]},${e.cell[1]}:${e.digit}`; if (seen.has(k)) return false; seen.add(k); return true;
      });
      if (!newElims.length) continue;
      const board = ctx.board;
      const carriersStr = [...m.carriers].sort((a,b)=>a[0]-b[0]||a[1]-b[1]).map(cellLabel).join(', ');
      const srcLbl = unitLabel(board.units[m.sourceUnitId]!);
      const tgtLbl = unitLabel(board.units[m.targetUnitId]!);
      const elimStr = [...newElims].sort((a,b)=>a.cell[0]-b.cell[0]||a.cell[1]-b.cell[1]).map(e=>cellLabel(e.cell)).join(', ');
      hints.push({
        ruleName: this.name,
        displayName: m.pattern === 'unit_cage' ? 'Locked Candidates (Cage-Line)' : 'Locked Candidates (Box-Line)',
        explanation: m.pattern === 'unit_cage'
          ? `In ${srcLbl}, ${m.digit} can only go in ${carriersStr}. All those cells belong to the same cage (${tgtLbl}). Since ${m.digit} must appear somewhere in ${srcLbl} and all its candidates are inside that cage, ${m.digit} can be eliminated from ${elimStr} (cage cells outside ${srcLbl}).`
          : `In ${srcLbl}, ${m.digit} can only go in ${carriersStr}. All those cells lie within ${tgtLbl}. Since ${m.digit} must appear somewhere in ${srcLbl} and all its candidates are locked to ${tgtLbl}, ${m.digit} can be eliminated from ${elimStr} (the other cells in ${tgtLbl} outside ${srcLbl}).`,
        highlightCells: [...m.carriers, ...newElims.map(e => e.cell)],
        eliminations: newElims,
        placement: null,
        virtualCageSuggestion: null,
      });
    }
    return hints;
  }
}
